src/checkout.c

```diff
@@ -1966,34 +1966,109 @@ static int conflict_entry_name(
 	return 0;
 }
 
+static int checkout_path_suffixed_0(
+	git_buf *path,
+	size_t basename_offset,
+	size_t basename_len,
+	const char *suffix,
+	size_t suffix_len,
+	int sequence) {
+	size_t basename_max = (NAME_MAX * 2) / 3;
+	size_t suffix_max = (NAME_MAX - basename_max) - 1;
+	size_t sequence_len;
+	char sequence_buf[(CHAR_BIT * sizeof(int) / 3) + 3];
+
+	/* Figure out the length of the suffix inclusive of the
+	 * sequence number */
+	if (sequence != -1) {
+		/* Would like to use snprintf here, but it's C99 */
+		sequence_len = 1 + sprintf(sequence_buf, "%d", sequence);
+		suffix_len += sequence_len;
+	}
+
+	if (basename_len + suffix_len + 1 > NAME_MAX) {
+		if (basename_len > basename_max) {
+			if (suffix_len > suffix_max) {
+				/* Both basename and suffix are too big, so
+				 * we allocate according to the (2/3, 1/3)
+				 * allocation.
+				 */
+				suffix_len = suffix_max;
+				basename_len = basename_max;
+			} else {
+				/* Just the basename is too long, so we can
+				 * steal a litle from the suffix's allocation.
+				 */
+				basename_len = NAME_MAX - suffix_len - 1;
+			}
+		} else {
+			/* Must be the suffix that's blowing the budget. */
+			suffix_len = NAME_MAX - basename_len - 1;
+		}
+	}
+
+	git_buf_truncate(path, basename_offset + basename_len);
+	if (sequence == -1) {
+		return git_buf_printf(path, "~%.*s", (int)suffix_len, suffix);
+	}
+	return git_buf_printf(path, "~%.*s_%d",
+			      (int)(suffix_len - sequence_len),
+			      suffix,
+			      sequence);
+}
+
 static int checkout_path_suffixed(git_buf *path, const char *suffix)
 {
-	size_t path_len;
-	int i = 0, error = 0;
+	int i = 0, error = 0, ret = 0;
+	size_t basename_len;
+	size_t suffix_len;
+	size_t basename_offset;
+	git_buf saved_path = GIT_BUF_INIT;
 
-	if ((error = git_buf_putc(path, '~')) < 0 || (error = git_buf_puts(path, suffix)) < 0)
-		return -1;
+	git_buf_put(&saved_path, git_buf_cstr(path), git_buf_len(path));
 
-	path_len = git_buf_len(path);
+	/* We need to ensure that the last component of the path (the
+	 * filename) is less than NAME_MAX, which is typically 255.
+	 * We're about to append a suffix (which might include some
+	 * digits)
+	 *
+	 * If our filename is already too long, we need to truncate
+	 * it.  We'll arbitrarily divide the available space as
+	 * follows: filename: 2/3; suffix and digits: 1/3 - 1 (for the
+	 * '~').  But if either filename or suffix doesn't completely
+	 * fill its allotted space, we allow the other to take up the
+	 * slack.
```

Comment:
Aren't these (at least partially) implementation details of `checkout_path_suffixed_0`?